doc/developer/design/20240411_graceful_reconfiguration_of_managed_clusters_v1.md

```diff
@@ -0,0 +1,355 @@
+performing an alter cluster on a managed cluster.
+
+We'll need to look at the alter_mechanism to determine if
+we need to run additional validations and whether to use the
+`sequence_alter_cluster_managed_to_managed_disruptive` or
+`sequence_alter_cluster_managed_to_managed_graceful` function.
+
+
+The following are validations to run if the `AlterClusterPlan` contains
+an alter_mechanism.
+1. validate the cluster is managed
+2. validate that the cluster is not undergoing a reconfiguration
+3. validate the cluster has no sources and now sinks
+4. the cluster cannot have a refresh schedule (v1)
+
+
+`sequence_alter_cluster_managed_to_managed_disruptive` will  perform the current
+delete-before-create mechanism.
+
+
+`sequence_alter_cluster_managed_to_managed_graceful`  will still apply
+all limits and validity checks to the `new configuration`, but will not
+perform deletes of the active replicas, nor commit the new config to the
+catalog. Instead, it will ask the compute controller to create new replicas,
+add a new entry in  `pending_cluster_reconfigurations`,  then kick off a
+`ClusterReconfigureCheckReady` message.
+
+The state for the reconfiguration will largely be stored in a `PendingClusterReconfiguration`
+struct.
+
+
+```rust
+struct PendingClusterReconfiguration {
+ ctx: ExecuteContext,
+ // used to determine if the message_handler
+ // should wait, or finalize
+ is_ready: bool
+ // needed to determine timeouts
+ start_time: Instant,
+ // the plan will hold the cluster_id, config, mechanism, etc...
+ plan: AlterClusterPlan,
+ // idempotency key will need to be part of the  message and will need to be
+ // validated to ensure we're looking at the correct plan, in case we
+ // miss a cancel and a new entry was put in pending_cluster_reconfiguration.
+ idempotency_key: i64,
+}
+
+```
+
+In order to cancel we'll keep a map, `pending_cluster_reconfigurations`,  of
+session `connectionIds` to `PendingClusterReconfigurations`.
+
+*Currently due to limitations in sources and sinks, we will need to maintain
+the disruptive alter mechanism. Once multiple replicas are allowed for sources
+and sinks we can remove the disruptive mechanism and default the wait to `WAIT=For 0 seconds`.
+
+
+#### Handling PendingClusterReconfigure Messages
+
+The responsibilities of this operation will be:
+1. checking for cancellations
+2. checking for the `WAIT` condition
+3. finalizaing the reocnfigure
+4. returning to the user
+
+
+__Finalize the Reconfiguration__
+Finalizing the reconfiguration will occur once the `wait` condition has been
+met. During this phase, the following steps will occur.
+- reconfiguration replica promotion to active
+- cluster configuration update to the new desired state
+- replicas with the previous configuration will be removed
+- replicas for the cluster will be removed from the `mz_pending_cluster_replicas` table
+
+3. Reverting the reconfiguration
+On failure or timeout of the reconfiguration, this task will revert to the state
```

Comment:
Reverting due to timeout is only a v2 feature right? Timeout in v1 should not revert, but proceed.